### PR TITLE
Fix ctrl z textfield

### PR DESCRIPTION
**Pull request by koppor**
Fixes https://github.com/JabRef/jabref/issues/11420

However, this is a hard one. Do we need to implement undo/redo for textfields for ourselves?

### Mandatory checks

- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by Siedlerchr** [comment]
This was working until recently. I wonder what happend to it. 

I would suggest adding a text field to the reproducer and try it out

**Comment by Siedlerchr** [comment]
ctrl + y still produces exception when I play around with multiple crtl+z and ctrl+y

```
javax.swing.undo.CannotRedoException
	at java.desktop/javax.swing.undo.UndoManager.tryUndoOrRedo(UndoManager.java:469)
	at java.desktop/javax.swing.undo.UndoManager.redo(UndoManager.java:453)
	at org.jabref@100.0.0/org.jabref.gui.undo.CountingUndoManager.redo(CountingUndoManager.java:37)
	at org.jabref@100.0.0/org.jabref.gui.undo.RedoAction.execute(RedoAction.java:21)
	at org.jabref@100.0.0/org.jabref.gui.actions.JabRefAction.lambda$new$1(JabRefAction.java:25)
	at org.controlsfx.controls@11.2.1/org.controlsfx.control.action.Action.handle(Action.java:423)
	at org.controlsfx.controls@11.2.1/org.controlsfx.control.action.Action.handle(Action.java:64)
	at javafx.base@22.0.1/com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
	at javafx.base@22.0.1/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:232)
	at javafx.base@22.0.1/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:189)
	at javafx.base@22.0.1/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at javafx.base@22.0.1/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base@22.0.1/com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at javafx.base@22.0.1/com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:49)
	at javafx.base@22.0.1/javafx.event.Event.fireEvent(Event.java:198)
	at javafx.controls@22.0.1/javafx.scene.control.MenuItem.fire(MenuItem.java:459)
	at javafx.controls@22.0.1/com.sun.javafx.scene.control.GlobalMenuAdapter.lambda$bindMenuItemProperties$2(GlobalMenuAdapter.java:150)
	at javafx.base@22.0.1/com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
	at javafx.base@22.0.1/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:232)
	at javafx.base@22.0.1/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:189)
	at javafx.base@22.0.1/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at javafx.base@22.0.1/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base@22.0.1/com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at javafx.base@22.0.1/com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
	at javafx.base@22.0.1/javafx.event.Event.fireEvent(Event.java:198)
	at javafx.controls@22.0.1/javafx.scene.control.MenuItem.fire(MenuItem.java:459)
	at javafx.graphics@22.0.1/com.sun.javafx.tk.quantum.GlassSystemMenu$1.action(GlassSystemMenu.java:237)

```

**Comment by koppor** [comment]
> ctrl + y still produces exception when I play around with multiple crtl+z and ctrl+y in comment field

This is not the original issue. The original issue is an NPE. - I will look into that nevertheless.

**Comment by koppor** [comment]
> javax.swing.undo.CannotRedoException
> 	at java.desktop/javax.swing.undo.UndoManager.tryUndoOrRedo(UndoManager.java:469)

Nice find! Was because of my OO refactoring. I did not copy the full branch for `RedoAction`. 😅

**Comment by koppor** [comment]
Follow up:

- Have menu items disabled if there is no undo / redo
- Implement undo/redo for "Clear, Change Case, Normalize"

**Comment by calixtus** [comment]
Please do not verschlimmbesser the undoredo stuff anymore.  Small refactorings ok, but the whole concept of undo/redo has to be redone.

**Comment by calixtus** [comment]
You can probably remove the now abstract undoredo class anyway and make the undo and the redo class directly extend simplecommand.

**Comment by koppor** [comment]
> You can probably remove the now abstract undoredo class anyway and make the undo and the redo class directly extend simplecommand.

Done in [`765e643` (#11421)](https://github.com/JabRef/jabref/pull/11421/commits/765e6435bfbc43b1dfefc0d5110b12e304baa982). I think, this code duplication is OK...

**Comment by koppor** [comment]
> * Have menu items disabled if there is no undo / redo

Discussed with @calixtus: Very hard. -- I also tried it locally, but the dependency of tab -> undomanger -> canUndo drives me crazy. I was searching for `Bindings.andThen`. Chain of bindings. But I did not find any solution.

**Comment by koppor** [comment]
>     * Implement undo/redo for "Clear, Change Case, Normalize"

I checked in this branch. Works.

**Comment by koppor** [comment]
@LoayGhreeb I reply to your comment https://github.com/JabRef/jabref/issues/11420#issuecomment-2198602298 here to be able to track the progress w.r.t. to the code commits. (The commits and the comments are sorted by time in a pull request. Thus, one sees, whether a comment was made before a commit or after one).

> Using the menu bar (Edit -> Undo) to undo changes works without any exceptions.

This is good 😅

> Another issue, which may or may not be related, is that when updating a field using the context menu (Clear, Change Case, Normalize) and using Ctrl+Z to undo this change, nothing happens and no exception occurs. 

I tested it with the latst commit here. Works.

> However, when using "Protect Selection" from the context menu, this change is undoable but with the exception.

Works here without exception.

Maybe, you can re-check? 😅

**Comment by LoayGhreeb** [comment]
> Maybe, you can re-check? 😅

I tried it, and it works without exceptions. However, can we do some checks before doing the undo/redo to see if the change is related to the focused field and selected entry or not?
Because now, if I add/remove an entry, pressing Ctrl + Z/Y in the text field will remove/restore the entry again.

**Comment by koppor** [comment]
>> Maybe, you can re-check? 😅
> I tried it, and it works without exceptions. However, can we do some checks before doing the undo/redo to see if the change is related to the focused field and selected entry or not?

Why? I was thinking that there should be one line of redo/undo changes and not some local textfield ones and some "global" ones not realted to the textfield ones?

I think, the behavior of the last release was that an undo of a text field (JavaFX behavior) was recorded as new change event in JabRef's global undo/redo. And then pressing undo from the menu lead to strange effects.

Now, there is only one global undo/redo chain. For sure, if one changes a text field, does somewhere else somethings, than goes back to the text field, types a letter, presses ctrl+Z twice, then the "somethings" are undone.

> Because now, if I add/remove an entry, pressing Ctrl + Z/Y in the text field will remove/restore the entry again.

This is the effect of a single global undo/redo stack.

Alternatively, we could "jump" in the undo chain and search for the last action regarding the text field and undo/redo that. -- I would stay back from using JavaFX textfield's local undo/redo (put aside for now that there is an NPE and we are in the need of fix JavaFX again).

File: src/main/java/org/jabref/gui/undo/CountingUndoManager.java [hunk 1]
```diff
@@ -4,39 +4,28 @@
...
-
-import com.google.common.eventbus.EventBus;
-
 public class CountingUndoManager extends UndoManager {
 
     private int unchangedPoint;
     private int current;
 
-    private final EventBus eventBus = new EventBus();
-
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/undo/CountingUndoManager.java:19]: 😍

**Review by calixtus — APPROVED** [review]

**Review by Siedlerchr — APPROVED** [review]

**Comment by calixtus** [comment]
> This is the effect of a single global undo/redo stack.
> 
> Alternatively, we could "jump" in the undo chain and search for the last action regarding the text field and undo/redo that. -- I would stay back from using JavaFX textfield's local undo/redo (put aside for now that there is an NPE and we are in the need of fix JavaFX again).

This can be resolved maybe by making undo actions in an event stream mergeable. So related edits (like entering a character) can be merged to larger undoables und be undone and redone as a compund (like now)

**Comment by koppor** [comment]
> > This is the effect of a single global undo/redo stack.
> > 
> > Alternatively, we could "jump" in the undo chain and search for the last action regarding the text field and undo/redo that. -- I would stay back from using JavaFX textfield's local undo/redo (put aside for now that there is an NPE and we are in the need of fix JavaFX again).
> 
> This can be resolved maybe by making undo actions in an event stream mergeable. So related edits (like entering a character) can be merged to larger undoables und be undone and redone as a compund (like now)

In addition, I meant:

1. Type a in field F1.
2. Type b in field F2.
3. Type c in Field F1.
4. Press Ctrl+Z twice in Field F1.

Loay expects F1 being empty.

Current behavior: F1 contains a and F2 contains nothing.

**Comment by calixtus** [comment]
i see.
i would expect the current behaviour. if you are jumping between fields, then every change has its own compund.

**Comment by koppor** [comment]
OK, Loay is asking for ["Objektbezogenes Undo"](http://www.usabilitypatterns.info/catalog/patterns/objektundo.html). The text there does not fully explain how it works with hierarchical objects. It also does not link real-world examples (which makes it a "pattern candidate")

**Comment by koppor** [comment]
> if you are jumping between fields, then every change has its own compund.

We need to craft this. As you outlined. - But maybe we should only append if it is the same operation: add / delete. Replace should be in a new thing. 

Meaning:

- (a, ab) + (ab, abc) ->(a, abc) // both changes were additions, merge second into first
- (abcd, abcde) + (abcde, abcd) -> (abcd, abcde), (abcde, abcd) // different actions, keep both, no merge

**Pull request merged**